lib/bandit/http1/adapter.ex

```diff
@@ -330,26 +306,24 @@ defmodule Bandit.HTTP1.Adapter do
   ##################
 
   @compile {:inline, read_available: 2}
-  @spec read_available(ThousandIsland.Socket.t(), timeout()) ::
+  @spec read_available(ThousandIsland.Socket.t(), keyword()) ::
           {:ok, binary()} | {:error, :closed | :timeout | :inet.posix()}
-  defp read_available(socket, read_timeout) do
-    ThousandIsland.Socket.recv(socket, 0, read_timeout)
+  defp read_available(socket, opts \\ []) do
+    ThousandIsland.Socket.recv(socket, 0, Keyword.get(opts, :read_timeout))
   end
 
-  @dialyzer {:no_improper_lists, read: 5}
-  @spec read(
-          ThousandIsland.Socket.t(),
-          non_neg_integer(),
-          iolist(),
-          non_neg_integer(),
-          timeout()
-        ) :: {:ok, iolist()} | {:error, :closed | :timeout | :inet.posix()}
-  defp read(socket, to_read, already_read, read_size, read_timeout) do
+  @dialyzer {:no_improper_lists, read: 4}
+  @spec read(ThousandIsland.Socket.t(), non_neg_integer(), iolist(), keyword()) ::
+          {:ok, iolist()} | {:error, :closed | :timeout | :inet.posix()}
+  defp read(socket, to_read, already_read, opts) do
+    read_size = Keyword.get(opts, :read_length, 1_000_000)
+    read_timeout = Keyword.get(opts, :read_timeout)
```

Comment:
Searching in a keyword list on every read call is not free. Passing timeouts in arguments (registers of VM) is virtually free.

Reply:
Very good point. I think the tradeoff here is that we only actually access the keywords when we *actually* read off the wire, vs before when we read them every time we called read_req_body. Moreover, we'll almost always be doing a handful of read calls (the default length to read is 1MB at a time). So I *think* this will actually net to fewer keyword calls.

Reply:
There are few counter arguments I could think of
* file uploads: upload of 100 megs equals to 200 keyword looks up which of course don't put that much overhead, but taking into account that there are only 2 places in module from where `read` is called, 200 calls cloud have been reduces just to 4.
* is :read_length configurable by a user? If so, someone might try reduce its value to smaller size to match for example their network stack configuration parameters (for some users latency is more important that throughput and they literally re-configuring the entire stack  to make their systems react faster in exchange some much they can process)

BTW is that `opts` coming from Plug dominion might change between calls? If no, why can't it be carried along with `THousandIsland.Socket` struct?

Reply:
The `opts` here is as passed via [`Plug.Conn.read_body`](https://hexdocs.pm/plug/Plug.Conn.html#read_body/2), so it varies from call to call.

Reply:
Fair point - I'll refactor them up to the point of first use, and get the best of both worlds.

Reply:
Updated

Reply:
Thank you. I believe that this thread is resolved :)